File: 2391.minimum-amount-of-time-to-collect-garbage.cs
/*
 * @lc app=leetcode id=2391 lang=csharp
 *
 * [2391] Minimum Amount of Time to Collect Garbage
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    public int GarbageCollection(string[] garbage, int[] travel) {
        Dictionary<char, List<int>> hashTable = new Dictionary<char, List<int>>() {
            { 'M', new List<int>() },
            { 'P', new List<int>() },
            { 'G', new List<int>() }
        };

        for (int i = 0; i < garbage.Length; i++) {
            foreach (char c in garbage[i]) {
                hashTable[c].Add(i);
            }
        }

        List<int> prefix = new List<int> { 0 };
        foreach (int t in travel) prefix.Add(prefix.Last() + t);

        int result = 0;

        foreach (List<int> route in hashTable.Values) {
            for (int i = 0; i < route.Count; i++) {
                result += 1 + prefix[route[i]] - prefix[i == 0 ? 0 : route[i - 1]];
            }
        }

        return result;
    }
}

// @lc code=end

File: 2391.minimum-amount-of-time-to-collect-garbage.cpp
/*
 * @lc app=leetcode id=2391 lang=cpp
 *
 * [2391] Minimum Amount of Time to Collect Garbage
 */

// @lc code=start
#include <vector>
#include <unordered_map>

class Solution {
public:
    int garbageCollection(std::vector<std::string>& garbage, std::vector<int>& travel) {
        std::unordered_map<char, std::vector<int>> hashTable = {
            { 'M', std::vector<int>() },
            { 'P', std::vector<int>() },
            { 'G', std::vector<int>() }
        };

        for (int i = 0; i < garbage.size(); i++) {
            for (char c : garbage[i]) {
                hashTable[c].push_back(i);
            }
        }

        std::vector<int> prefix(1, 0);
        prefix.insert(prefix.end(), travel.begin(), travel.end());

        for (int i = 1; i < prefix.size(); i++) {
            prefix[i] += prefix[i - 1];
        }

        int result = 0;

        for (const auto& route : hashTable) {
            for (int i = 0; i < route.second.size(); i++) {
                result += 1 + prefix[route.second[i]] - prefix[i == 0 ? 0 : route.second[i - 1]];
            }
        }

        return result;
    }
};

// @lc code=end

